### PR TITLE
[WIP] Add structured result output

### PR DESCRIPTION
**Pull request by williamFalcon**
This PR maintains full backward compatibility, but for those who want the option adds an optional argument to the *_step that is a structured dict with validation for managing things to return from each step

Old way (still supported)

```python
def training_step(self, batch, batch_idx...)
    return {...}
```

New way

```python
  # any loop
  def training_step(self, batch, batch_idx...)
        """
        Lightning calls this inside the training loop with the data from the training dataloader
        passed in as `batch`.
        """
        # forward pass
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)

        # structure the return from the training loop
        return Result(loss)
```

Docs:
```python
            # all options:
            def training_step(...):
                return Result(
                    minimize=loss,
                    checkpoint_on=loss,
                    early_stop_on=loss,
                    logs={'train_loss': loss},
                    progress_bar={'train_loss': loss}
                )

            # most of the time
            # will early stop and save checkpoints based on this metric by default
            return Result(loss)

            # to change what to early stop on
            return Result(loss, early_stop_on=accuracy)

            # to change what to checkpoint on
            return Result(loss, early_stop_on=accuracy, checkpoint_on=bleu_score)

            # shorthand for logging
            result = Result(loss)
            result.log('train_nce_loss', loss)

            # shorthand to put on progress bar
            result.to_bar('train_nce_loss', loss)
```

Additional benefits:
- gets rid of 'val_loss' magic. Now user can explicitly set what to stop or save on.
- gets rid of 'loss' now user can call it whatever they want (we're just minimizing it).
- adds error checking if the user puts in the wrong details.
- clear separation of what each return item does

### CODE REVIEW

**Comment by williamFalcon** [comment]
@tullie is this in line with what you were thinking?
Will ping you when done

**Comment by tullie** [comment]
Why does StepResult need to be passed in as a training_step argument? That seems unintuitive to me.

Could we aim for something like:
```
def training_step(self, batch: Tensor, batch_idx: int):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        return StepResult(loss=loss, log=loss)
```

**Comment by williamFalcon** [comment]
@tullie ok, updated. check out the new API before i finish making all the deep changes

**Comment by jeremyjordan** [comment]
related to #1256, happy to see progress here!

**Comment by tullie** [comment]
@williamFalcon yeah this is great. Huge improvement imo! I'll let you finish the TODOs and then do a closer sweep of all the code.

**Comment by Borda** [comment]
> as i mentioned in the other issue, it might be nice to use something like `pydantic` for the structured output since it provides very easy data validation for free

this looks very similar to `Namespace` except the validation and other features around... :]

**Comment by williamFalcon** [comment]
@jeremyjordan #1256 is awesome, must have missed that haha. Why don't i put together the v1 right now and you can take a stab at V2? happy to make this a joint PR. with this object we can now do whatever we want under the hood for validation :)

**Comment by jeremyjordan** [comment]
yeah, sounds great!

**Comment by Borda** [comment]
is it still WIP? :raccoon:

File: pl_examples/models/simple_template.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        """
+        return optim.Adam(self.parameters(), lr=self.learning_rate)
+
+    @staticmethod
+    def add_model_specific_args(parent_parser):  # pragma: no-cover
+        """
+        Define parameters that only apply to this model
+        """
+        parser = ArgumentParser(parents=[parent_parser], add_help=False)
+        parser.add_argument('--in_features', default=28 * 28, type=int)
```
Comment (Borda) [inline pl_examples/models/simple_template.py:125]: this can be parsed automatically from Models units arguments...

**Comment by williamFalcon** [comment]
yes. but feel free to comment on the current api

File: pytorch_lightning/core/step_result.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
+from typing import Optional, Dict
+from torch import Tensor
+import torch
+
+
+class Result(Dict):
+
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:7]: adding docstring and doctest here would be great, I can do it later on

File: pytorch_lightning/core/step_result.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+            assert isinstance(x, Tensor), 'hiddens must be a torch.Tensor'
+            self._hiddens = x
+            self.__setitem__('hiddens', x)
+
+    @property
+    def checkpoint_on(self):
+        # use minimize as default if no checkpoint_on is passed
+        if 'checkpoint_on' not in self:
+            minimize = self.__getitem__('minimize')
+            self.__setitem__('checkpoint_on', minimize)
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:170]:
```suggestion
            self.__setitem__('checkpoint_on', self.minimize)
```

File: pytorch_lightning/core/step_result.py [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+        if x is not None:
+            assert isinstance(x, Tensor), 'checkpoint_on must be a torch.Tensor'
+            self.__setitem__('checkpoint_on', x.detach())
+
+    @property
+    def early_stop_on(self):
+        # use minimize as default if no checkpoint_on is passed
+        if 'early_stop_on' not in self:
+            minimize = self.__getitem__('minimize')
+            self.__setitem__('early_stop_on', minimize)
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:185]:
```suggestion
            self.__setitem__('early_stop_on', self.minimize)
```

File: pytorch_lightning/core/step_result.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    def to_pbar(self, name: str, value: Tensor, on_batch_end=False, on_epoch_end=True, reduce_fx=torch.mean):
+        # no graph pointers for progress bar
+        value = value.detach()
+        super().to_pbar(name, value, on_batch_end, on_epoch_end, reduce_fx)
+
+
+if __name__ == '__main__':
+    import torch
+    result = Result()
+    result.minimize = torch.tensor(1)
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:382]:
```suggestion
```
this may go to doctest...

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -17,8 +17,103 @@
...
+    assert out.log_on_epoch_end['log_acc2'] == 7.0
+    assert out.pbar_on_epoch_end['pbar_acc1'] == 17.0
+    assert out.pbar_on_epoch_end['pbar_acc2'] == 19.0
+    assert out['early_stop_on'] == 1.4
+    assert out['checkpoint_on'] == 1.5
+
+    # ---------------------
+    # test dic return only
+    # ---------------------
+    model = DeterministicModel()
```
Comment (Borda) [inline tests/trainer/test_trainer.py:50]: this looks the same as the above...

File: pytorch_lightning/core/step_result.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+        if on_epoch_end:
+            self.__reduce_on_callback('on_epoch_end', name, value, log=True, pbar=False, reduce_fx=reduce_fx)
+
+    @property
+    def log_on_batch_end(self):
+        return self.__getitem__('log_on_batch_end')
+
+    @log_on_batch_end.setter
+    def log_on_batch_end(self, x):
+        if x is not None:
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:80]: what is the case you would pass NOne here or any very same setter?

File: pytorch_lightning/core/step_result.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+                 checkpoint_on: Tensor = None,
+                 hiddens: Optional[Tensor] = None):
+        super().__init__()
+
+        self.early_stop_on = early_stop_on
+        self.checkpoint_on = checkpoint_on
+
+        # TODO: should hiddens detach?
+        self.hiddens = hiddens
+        self.minimize = minimize
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:20]: is there a point to have it here as attribute and alter as property?

File: pytorch_lightning/core/step_result.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+            self[key] = {}
+
+        metrics = self[key]
+        metrics[name] = reduce_fx
+
+    def pass_to_epoch_end(self, name, metric):
+        if 'to_epoch_end' not in self:
+            self['to_epoch_end'] = {}
+
+        metrics = self['to_epoch_end']
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:52]: still think that having internal `metrics` and in `progress_bar` just save a list of metrics keys (on-demand you would dictionary as you can simply zip it in the property) which may reduce user confusion that log metric and progress bar is different because we accidentally did not update one of them...

File: pytorch_lightning/utilities/parsing.py [hunk 10]
```diff
@@ -47,3 +47,8 @@ def clean_namespace(hparams):
 
         for k in del_attrs:
             del hparams[k]
+
+
+class AttributeDict(dict):
```
Comment (williamFalcon) [inline pytorch_lightning/utilities/parsing.py:52]: @Borda here

**Comment by import-antigravity** [comment]
Is this still in active development? It would be really nice to have :)

**Comment by williamFalcon** [comment]
yes!! in #2615
